src/sympc/protocol/aby3/aby3.py

```diff
@@ -0,0 +1,161 @@
+"""ABY3 Protocol.
+
+ABY3 : A Mixed Protocol Framework for Machine Learning.
+https://eprint.iacr.org/2018/403.pdf
+"""
+# stdlib
+from typing import Any
+from typing import List
+from typing import Tuple
+
+# third party
+import torch
+import torchcsprng as csprng
+
+from sympc.protocol.protocol import Protocol
+from sympc.session import Session
+from sympc.tensor import MPCTensor
+from sympc.tensor import ReplicatedSharedTensor
+from sympc.tensor.tensor import SyMPCTensor
+
+gen = csprng.create_random_device_generator()
+
+
+class ABY3(metaclass=Protocol):
+    """ABY3 Protocol Implementation."""
+
+    """Used for Share Level static operations like distributing the shares."""
+    share_class: SyMPCTensor = ReplicatedSharedTensor
+    security_levels: List[str] = ["semi-honest", "malicious"]
+
+    def __init__(self, security_type: str = "semi-honest"):
+        """Initialization of the Protocol.
+
+        Args:
+            security_type : specifies the security level of the Protocol.
+
+        Raises:
+            ValueError : If invalid security_type is provided.
+        """
+        if security_type not in self.security_levels:
+            raise ValueError(f"{security_type} is not a valid security type")
+
+        self.security_type = security_type
+
+    def __eq__(self, other: Any) -> bool:
+        """Check if "self" is equal with another object given a set of attributes to compare.
+
+        Args:
+            other (Any): Object to compare
+
+        Returns:
+            bool: True if equal False if not.
+        """
+        if self.security_type != other.security_type:
+            return False
+
+        if type(self) != type(other):
+            return False
+
+        return True
+
+    @staticmethod
+    def trunc1(
+        ptr_list: List["ReplicatedSharedTensor"], shape: torch.Size, session: Session
+    ) -> List["ReplicatedSharedTensor"]:
+        """Performs the ABY3 trunc1 algorithm.
+
+        Args:
+            ptr_list (List[ReplicatedSharedTensor]): Tensors to truncate
+            shape(torch.Size) : shape of tensor values
+            session(Session) : session the tensor belong to
+
+        Returns:
+            List["ReplicatedSharedTensor"] : Truncated shares.
+        """
+        rand_value = torch.empty(size=shape, dtype=session.tensor_type).random_(
+            generator=gen
+        )
+        base = session.config.encoder_base
+        precision = session.config.encoder_precision
+        scale = base ** precision
+        x1 = ptr_list[0].get_copy().shares[0]
+        x2, x3 = ptr_list[1].get_copy().shares
+        x1_trunc = x1 >> precision if base == 2 else x1 // scale
+        x_trunc = (x2 + x3) >> precision if base == 2 else (x2 + x3) // scale
+        shares = [x1_trunc, x_trunc - rand_value, rand_value]
+        ptr_list = ReplicatedSharedTensor.distribute_shares(shares, session)
```

Comment:
You could keep this. But, what we could do is remove resharing in multiplication. Truncation requires 3 values. We could use z1,z2&z3 after Falcon mult for truncation and then perform 2-3 resharing rather than do it twice?

Reply:
Yeah thought would be communication efficient, Sure I would change it.

Reply:
We do share generation at distribution ,so I implemented in this manner.